### ISSUE TITLE
Some packages will still build despite the same version being on the binary repo

### ISSUE THREAD

**Issue by MoreRobustThanYou** [postmarketOS/pmbootstrap#955]
Examples: ```armhf/breeze-icons-5.40.0-r0.apk```,  ```x86_64/binutils-armhf-2.28-r3.apk```, ```armhf/kcoreaddons-5.40.0-r0.apk```

I did pmbootstrap zap, update, then index, but it STILL tries to locally build these rather than fetching them from binary repo. Ubuntu 17.10, pmbootstrap up to date with master.

**Comment by ollieparanoid** [comment]
Thanks for your report! Could you please try the following?

```shell
# delete the breeze-icons package to reproduce the issue
sudo rm ~/.local/var/pmbootstrap/packages/armhf/breeze-icons-5.40.0-r0.apk
pmbootstrap index

# zap and update APKINDEX files
pmbootstrap -y zap
pmbootstrap update

# clear the log file (~/.local/var/pmbootstrap/log.txt, make a backup if you need it)
pmbootstrap log -c

# build it with verbose logging. hopefully this reproduces the same issue
# (-> it builds the package and does not download it!)
pmbootstrap -v build breeze-icons

# please attach this logfile
cat ~/.local/var/pmbootstrap/log.txt
```

**Comment by JLIT0** [comment]
I had a similar issue on Manjaro where pmbootstrap tried to build the entire system from scratch (at step 2/5) instead of using the binary files.

Strangely, this issue didn't exist on Ubuntu-based distros or Fedora, but these both had their own issues in that graphical acceleration in QEMU wasn't supported.  I could only get to a Weston screen by specifying `gl=off` in both.  (Hildon and XFCE failed to open the framebuffer without graphics acceleration, as expected.)  In the case of Ubuntu, I believe this is due to #906 , and in Fedora due to some sort of OpenGL issue related to `libvirt`.

**Comment by ollieparanoid** [comment]
@JLIT0: Thanks for reporting! If you could provide the verbose log (as written above) for the packages which get compiled instead of downloaded, that would be equally helpful.

All Qemu errors are worked around by using the LTS kernel for now, so that shouldn't be an issue anymore (that was just merged yesterday in #954).

**Comment by zhuowei** [comment]
Not sure if I'm running into the same issue with qt5-qtsensors 5.9.3-r0.

It seems that the check looks at the local repo first, and it's picking up my locally compiled 5.9.1-r0 and concluding that the package needs to be recompiled, without checking the online binary repo.

```
(018516) [12:02:34] *** (1/5) PREPARE NATIVE CHROOT ***
(018516) [12:02:34] Calculate depends of packages ['cryptsetup', 'util-linux', 'e2fsprogs', 'parted'], arch: x86_64
(018516) [12:02:35] *** (2/5) CREATE DEVICE ROOTFS ("qemu-amd64") ***
(018516) [12:02:35] (rootfs_qemu-amd64) % apk update
fetch http://postmarketos.brixit.nl/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/edge/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/edge/community/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/edge/testing/x86_64/APKINDEX.tar.gz
pmos  [/mnt/pmbootstrap-packages]
pmos  [http://postmarketos.brixit.nl]
v3.7.0-196-ga8b508cc5f [http://dl-cdn.alpinelinux.org/alpine/edge/main]
v3.7.0-200-g52c914aa8a [http://dl-cdn.alpinelinux.org/alpine/edge/community]
v3.7.0-199-g0da55331ad [http://dl-cdn.alpinelinux.org/alpine/edge/testing]
OK: 12401 distinct packages available
(018516) [12:02:36] Calculate depends of packages dict_keys(['zlib', 'cmd:lbu_status', 'cmd:lbu_commit', 'cmd:ldconfig', 'cmd:setup-gparted-desktop', 'apk-tools', 'command:mkmntdirs', 'cmd:setup-disk', 'cmd:getconf', 'cmd:setup-acf', 'so:libc.musl-x86_64.so.1', 'cmd:lbu_update', 'so:librc.so.1', 'libressl2.6-libssl', 'cmd:setup-dns', 'cmd:setup-apkrepos', 'cmd:bbsuid', 'cmd:setup-apkcache', 'cmd:rc-status', 'busybox-suid', 'so:libeinfo.so.1', 'cmd:setup-timezone', 'alpine-base', 'so:libz.so.1', 'cmd:start-stop-daemon', 'musl-utils', 'alpine-conf', 'busybox', 'cmd:setup-alpine', 'cmd:lbu_include', 'cmd:service', 'libc-utils', 'cmd:iconv', 'cmd:setup-hostname', 'alpine-baselayout', 'so:libcrypto.so.42', 'libressl2.6-libcrypto', 'cmd:setup-keymap', 'cmd:lbu_exclude', 'musl', 'cmd:setup-ntp', 'cmd:lbu', 'cmd:runscript', 'cmd:openrc-run', 'alpine-keys', 'cmd:busybox', 'openrc', 'busybox-initscripts', 'cmd:setup-mta', 'cmd:rc-sstat', 'cmd:rc-update', '/bin/sh', 'cmd:openrc', 'cmd:setup-lbu', 'cmd:setup-xen-dom0', 'cmd:supervise-daemon', 'cmd:setup-proxy', 'cmd:update-kernel', 'cmd:setup-xorg-base', 'so:libssl.so.44', 'cmd:setup-interfaces', 'cmd:rc', 'cmd:uniso', 'cmd:setup-bootable', 'cmd:setup-sshd', 'cmd:update-conf', 'cmd:sh', 'scanelf', 'cmd:apk', 'cmd:rc-service', 'cmd:getent']), arch: x86_64
(018516) [12:02:36] (rootfs_qemu-amd64) % getent passwd user
user:x:1000:1000:Linux User,,,:/home/user:/bin/ash
(018516) [12:02:37] Build is necessary for package 'qt5-qtsensors': Binary package out of date (binary: 5.9.1-r0, aport: 5.9.3-r0)
(018516) [12:02:37] Calculate depends of packages ['qt5-qtbase', 'qt5-qtbase-dev'], arch: x86_64
(018516) [12:02:38] (native) build x86_64/qt5-qtsensors-5.9.3-r0.apk
```

I logged the output values of the apkindex.parse function:

```
retver /media/zhuowei/redhd/docs/amd64_pmbootstrap/packages/x86_64/APKINDEX.tar.gz 5.9.1-r0
qtsensors  5.9.3-r0 5.9.1-r0 1
retver /media/zhuowei/redhd/docs/amd64_pmbootstrap/cache_apk_x86_64/APKINDEX.8b865e19.tar.gz 5.9.3-r0
[12:17:34] *** (2/5) CREATE DEVICE ROOTFS ("qemu-amd64") ***
retver /media/zhuowei/redhd/docs/amd64_pmbootstrap/cache_apk_x86_64/APKINDEX.30e6f5af.tar.gz 5.9.3-r0
5.9.1-r0
[12:17:37] (native) build x86_64/qt5-qtsensors-5.9.3-r0.apk
```

I think apkindex.read_any_index should be changed to return the package with the highest version number instead of the first package found.

**Comment by ollieparanoid** [comment]
Awesome, thanks for the analysis! That makes a lot of sense.

**Comment by ollieparanoid** [comment]
I've made a fix in #1004. Reviewing and testing from anyone appreciated.

**Comment by ollieparanoid** [comment]
Turns out the fix is only for what @zhuowei described (reviews/testing welcome!), so I've created another issue to further debug @MoreRobustThanYou's original issue in #1023.
